### ISSUE TITLE
deallocation on final

### ISSUE THREAD

**Issue by kookma** [tomedunn/fortran-linked-list#1]
The list uses allocation in "list_append_item" subroutine, and there is no deallocation in "final" destructor.
Isn't there any need to deallocate the list items? Have you measured the memory usage during run if many times the list used/reused?

**Comment by zbeekman** [comment]
Yes, to me it looks like this will leak memory. Pointers are correctly nullified, but the actual objects are not deallocated by the destructors.

**Comment by tomedunn** [comment]
I haven't done any sort of performance or memory leak testing on this code. But I would trust @zbeekman on this. He has a lot more experience I'm sure using practical code that uses these features than I do. I wrote this as a toy model back when some of these things, while part of the Fortran standard, were barely supported by the compilers available at the time. So my main concern was does it work or not, not how well does it work.

**Comment by zbeekman** [comment]
You *may* be able to replace a bunch of the pointers (i.e. the node type) with allocatable components w/ target attributes for modern enough F2008/F2018 compilers. Would be a fun test. I'm guessing you may hit lots of nasty compiler bugs if your compiler claims to support unlimited polymorphic allocatables, but I would be happy to be wrong.

**Comment by kookma** [comment]
Thank you both!
As @zbeekman said, there is a memory leak. I used the code with a loop of 1 000 000 time trials. In each iteration a list is created and then destroyed. I used the Performance monitor in Windows with a Core i7 + 8 GB DDR4 machine. The memory usage starts to increase and it continue to full memory usage.

I appreciate if you add a line in the readme.md and note that, this code has memory leak and use it with care. By the way thank you for your effort and sharing your Fortran experiences.

**Comment by zbeekman** [comment]
Leak is pretty trivial to fix, so long as you disallow things to point at the list/list elements when they go out of scope.

**Comment by kookma** [comment]
Thank you @zbeekman 
I used the concept in another code and I take care of memory leak!
By the way, if the linked list is used for a type contains an allocatable member which is in turn has some allocatable members itself, do you think it is necessary to allocate all those members one by one?

**Comment by zbeekman** [comment]
All members will need to be individually allocated, under many circumstances. Reallocation on assignment might fix some of this for you though. If you have

``` Fortran
  type :: node
    private
    type(node), allocatable, target :: next
    class(*), allocatable :: item
  end type node
```

Each `...%next` element will need to be allocated.

However, when one or more go out of scope all the components should be automatically deallocated by the compiler, starting with the tail element.
